peps/pep-9995.rst

```diff
@@ -0,0 +1,228 @@
+PEP: 9995
+Title: Inlined typed dictionaries
+Author: Victorien Plot <contact@vctrn.dev>
+Discussions-To:
+Status: Draft
+Type: Standards Track
+Topic: Typing
+Created: 23-Oct-2024
+Post-History:
+Python-Version: 3.14
+Resolution:
+
+.. highlight:: python
+
+
+Abstract
+========
+
+:pep:`589` defines a `class-based <typing:typeddict-class-based-syntax>`_ and a
+:ref:`functional syntax <typing:typeddict-functional-syntax>` to create typed
+dictionaries. In both scenarios, it requires defining a class or assigning to
+a value. In some situations, this can add unnecessary boilerplate, especially
+if the typed dictionary is only used once.
+
+This PEP proposes the addition of a new inlined syntax, by subscripting the
+:class:`~typing.TypedDict` type::
+
+    from typing import TypedDict
+
+    def get_movie() -> TypedDict[{'name': str, 'year': int}]:
+        return {
+            'name': 'Blade Runner',
+            'year': 1982,
+        }
+
+Motivation
+==========
+
+Python dictionaries are an essential data structure of the language. Many
+times, it is used to return or accept structured data in functions. However,
+it can get tedious to define :class:`~typing.TypedDict` classes:
+
+* A typed dictionary requires a name, which might not be relevant.
+* Nested dictionaries requires more than one class definition.
+
+Taking a simple function returning some nested structured data as an example::
+
+    from typing import TypedDict
+
+    class ProductionCompany(TypedDict):
+        name: str
+        location: str
+
+    class Movie(TypedDict):
+        name: str
+        year: int
+        production: ProductionCompany
+
+
+    def get_movie() -> Movie:
+        return {
+            'name': 'Blade Runner',
+            'year': 1982,
+            'production': {
+                'name': 'Warner Bros.',
+                'location': 'California',
+            }
+        }
+
+
+Rationale
+=========
+
+The new inlined syntax can be used to resolve these problems::
+
+    def get_movie() -> TypedDict[{'name': str, year: int, 'production': {'name': str, 'location': str}}]:
```

Comment:
I don't think we should support nested inlined TypedDict definitions as shown above. We should require that the type annotation for each item be a valid type expression, and a dictionary expression not enclosed by a `TypedDict` is an invalid type expression. The above example should therefore be changed to:

```python
def get_movie() -> TypedDict[{"name": str, "year": int, "production": TypedDict[{"name": str, "location": str}]}]: ...
```

The spec should make it clear that using nested dictionary expressions (like in the current code sample) is not allowed and should result in a type checker error.

I understand this might seem needlessly verbose, but if we allow dictionary expressions in type expressions outside of a `TyepdDict` type argument, it will create a cascade of exceptions and future composability issues. For example, we'd need to support raw dictionary expressions within `Required`, `NotRequired`, and `ReadOnly`.

Reply:
Indeed using a nested inlined typed dictionary is more consistent! Changed.